### PR TITLE
User/system replica IDs

### PR DESCRIPTION
**Pull request by teskje**
In #18376 we introduced a new replica ID format that made it possible to differentiate between user and system replicas, like with clusters. That PR only made the minimal changes required to change the format in the system tables. This PR now makes the database code use the new replica ID format everywhere. It also changes the IDs of replicas within system clusters to system IDs.

The externally visible changes from this PR are:

1. Orchestrator service names now contain `s{n}`/`u{n}` replica IDs, rather than just `{n}`.
2. The format of replica IDs in `mz_audit_events` has changed.
3. System replicas now have `s{n}` IDs, rather than `u{n}` IDs.

I expect (1) and (2) to potentially require changes to the cloud code, which is why I added the blocker below.
[Relevant Slack thread.](https://materializeinc.slack.com/archives/CL68GT3AT/p1683901234330229)

### Blockers

* [ ] Test with cloud code and adjust if necessary. 

### Motivation

  * This PR adds a known-desirable feature.

Advances #18377.

### Tips for reviewer

Look at the commits and their descriptions separately!

I think most of the changes are straightforward, except maybe the catalog migration stuff. I initially tried to write a regular migration but [found out that that is not possible](https://materializeinc.slack.com/archives/C02FWJ94HME/p1681464979565619). @parker-timmerman pointed me to https://github.com/MaterializeInc/materialize/pull/18719, in which he makes a similar change using a backwards-compatible serialized format. However, I decided to use a different approach following https://github.com/MaterializeInc/materialize/pull/18719#discussion_r1166010915. This way, we won't end up with tech debt once the non-standard migration has been removed.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - N/A

### CODE REVIEW

File: src/adapter/src/catalog.rs [hunk 1]
```diff
@@ -4007,14 +4007,14 @@ impl Catalog {
         self.storage().await.get_all_persisted_timestamps().await
     }
 
-    /// Get the next user ID without allocating it.
-    pub async fn get_next_user_global_id(&self) -> Result<GlobalId, Error> {
-        self.storage().await.get_next_user_global_id().await
+    /// Get the next system replica id without allocating it.
+    pub async fn get_next_system_replica_id(&self) -> Result<u64, Error> {
+        self.storage().await.get_next_system_replica_id().await
     }
```
Comment (teskje) [inline src/adapter/src/catalog.rs:4013]: I removed the `get_next_user_global_id` method because it isn't used anywhere.

**Comment by necaris** [comment]
Tagging Cloud billing-knowledgeable folks (i.e. @necaris @arusahni ) to ensure that a companion Cloud change is made to keep billing from audit events consistent. Tentative plan is to update the audit events SQL queries to homogenize replica IDs.

**Comment by benesch** [comment]
Recording a proposal from Slack:

> I have a third proposal to offer: write a “migration” that emits drop events to the audit log for all existing replicas using old format IDs and then emits create events for the same replicas with the new format ID. That will ensure that every replica create event has a drop event with the same ID format.

We'd want to emit these events as user `mz_system`, I think. AFAICT, this migration would require no changes to cloud code, would avoid rewriting history in the audit log, and would ensure that the audit log uses the new-format IDs going forward.

I'm not sure this is better than @mjibson and @necaris's previous proposal of rewriting the (observable) audit log (though not the on-disk audit log). But wanted to throw it out.

**Comment by mjibson** [comment]
> drop events to the audit log for all existing replicas using old format IDs and then emits create events

If billing queries depend on this having happened, then there will be data in the audit log the billing queries cannot read, right? I suspect whatever solution we come up with here will be used when this class of problem happens again. If so, this is a bad precedent to start.

A point I've made many times (one was many months ago in maybe the cloud repo, but I'm unable to find it): the audit log is currently designed to be able to do this but we chose to not use that feature. One of its goals was to prevent exactly this problem by forcing users of the data to explicitly handle all versions. But because we're copying out the raw JSON from the stash into the audit events table, we're unable to get any type safety across versioned events. Teaching the audit events table how to do this instead of blind copying the JSON sets a very safe precedent (if annoying, an acceptable price to pay for getting billing queries correct though) for this and future type changes we want to make.

**Comment by benesch** [comment]
> If billing queries depend on this having happened, then there will be data in the audit log the billing queries cannot read, right?

The billing queries won't care; the ID is just an opaque string. All the billing queries care about is that the string used in the drop event exactly matches the string used in the create event. The proposed migration is just to make things "nicer" if you're only looking at recent history.

> A point I've made many times (one was many months ago in maybe the cloud repo, but I'm unable to find it): the audit log is currently designed to be able to do this but we chose to not use that feature. One of its goals was to prevent exactly this problem by forcing users of the data to explicitly handle all versions. But because we're copying out the raw JSON from the stash into the audit events table, we're unable to get any type safety across versioned events. Teaching the audit events table how to do this instead of blind copying the JSON sets a very safe precedent (if annoying, an acceptable price to pay for getting billing queries correct though) for this and future type changes we want to make.

I don't follow. The `mz_audit_events` table is part of Materialize's stable interface. I think the question at hand is whether it is permissible for a row in that table that was previously

```
4  create  cluster-replica  {"cluster_id":"u1","cluster_name":"default","logical_size":"1","replica_id":"1","replica_name":"r1"}  NULL
```

to suddenly become:

```
4  create  cluster-replica  {"cluster_id":"u1","cluster_name":"default","logical_size":"1","replica_id":"u1","replica_name":"r1"}  NULL
```

IMO, no, that's a BC break; users may previously have exported the old event with `replica_id: 1` and suddenly changing that event to `replica_id: u1` is very sketchy.

**Comment by teskje** [comment]
> I have a third proposal to offer: write a “migration” that emits drop events to the audit log for all existing replicas using old format IDs and then emits create events for the same replicas with the new format ID. That will ensure that every replica create event has a drop event with the same ID format.

I added a new commit that does that migrations. Depending on how the discussion pans out, we can keep it or drop it again. I'm lacking the context to say whether this is the best solution, but it seems to be the simplest, provided that it indeed doesn't break billing.

**Comment by mjibson** [comment]
> The mz_audit_events table is part of Materialize's stable interface.

This PR appears to be changing `mz_cluster_replica_history` and `mz_clusters`. Are those part of the stable interface? I agree that changing mz_audit_events is breaking, but I'm not sure it's more or less breaking than changing those other objects too. Making fake events to rename things also has oddities.

I'm fine moving forward with the PR as is.

**Comment by benesch** [comment]
> This PR appears to be changing `mz_cluster_replica_history` and `mz_clusters`. Are those part of the stable interface?

`mz_cluster_replica_history` is in `mz_internal` and so not part of the stable interface. `mz_cluster_replicas` _did_ have a backwards incompatible change as a result of earlier work, but we specifically made an exception for that change and very carefully ensured that it wouldn't brick any environments in production: https://materializeinc.slack.com/archives/CTESPM7FU/p1679906648029879

> I agree that changing mz_audit_events is breaking, but I'm not sure it's more or less breaking than changing those other objects too. Making fake events to rename things also has oddities.

I agree. Sorry, I think I haven't been communicating super clearly! Let me try again:

  1. We have to do _something_ funky to update the audit log for the backwards incompatible change to replica IDs. We've already agreed that changing the replica IDs in this backwards incompatible way is worth doing, so now it's just a question of what do we do with the old format IDs in the audit log. I'm truly not sure which of the migration strategies (synthesize a drop/create event for each extant replica, or update historical create events to use the new format IDs) is better.
  2. For _future_ changes we want to make to audit events, I'm not sure how the audit events crate will make the situation better for end users, because the type of the `details` field is `jsonb`, which means end users won't get any type safety. Whether the event is a "v1" event or "v2" event is lost by the time the data hits the SQL table. Maybe I'm misunderstanding your proposal in your last comment?

**Comment by mjibson** [comment]
> Whether the event is a "v1" event or "v2" event is lost by the time the data hits the SQL table. Maybe I'm misunderstanding your proposal in your last comment?

My proposal is that we add methods to the `EventDetails` enum that are called based on the `(event_type, object_type)` of each VersionedEvent, and those are the things put into `audit_log_events`. This is designed to abstract away from which event version is there but have a common (although not stable) interface over all versions. It's big downside as you point out is it can change between versions.

The solution to issue drop/create seems like a good solution. I'm ok with this.

**Comment by necaris** [comment]
It looks like as of now the drop / create solution will not affect the billing queries and ongoing reporting, so I 👍🏽

**Review by necaris — APPROVED** [review]

**Comment by teskje** [comment]
Okay, tests are passing, so I'm marking this ready for review. I also deployed on my staging account and the invoice looks fine so far, will check again tomorrow in case it needs a while to update.

@necaris There is one thing in cloud that we still need to fix. This line:
https://github.com/MaterializeInc/cloud/blob/a3e7eb5812f1d46fe906999df5e71770e17021ba/infra/cli/run_k8s_tests.py#L377
... needs to be made smarter to use the new service name for new MZ versions. Is there a way to get the MZ version somehow at this point? Perhaps from the `environment_assignment` dict?

**Comment by teskje** [comment]
@philip-stoev Let me know if you think more tests are in order. I could probably add an upgrade test that verifies that the new replica ID appears in the audit log after an upgrade.

**Comment by philip-stoev** [comment]
Item No1 . No audit events are emitted for the migration of the system replicas when upgrading from a previous version:

This is the state of the system tables post-upgrade:

```
materialize=> select * from mz_cluster_replicas;
 id |   name   | cluster_id | size | availability_zone | owner_id 
----+----------+------------+------+-------------------+----------
 s1 | r1       | s1         | 1    |                   | s1
...
(5 rows)

materialize=> select * from mz_audit_events where mz_audit_events.details->>'replica_id' = 's1';
 id | event_type | object_type | details | user | occurred_at 
----+------------+-------------+---------+------+-------------
(0 rows)
```

At the same time, there is a create+drop audit event for the creation of that same replica prior to the migration:

```
id | event_type |   object_type   |                                                details                                                 | user |        occurred_at         
----+------------+-----------------+--------------------------------------------------------------------------------------------------------+------+----------------------------
 4  | create     | cluster-replica | {"cluster_id":"u1","cluster_name":"default","logical_size":"4-4","replica_id":"1","replica_name":"r1"} |      | 2023-05-18 09:03:38.294+00
 12 | drop       | cluster-replica | {"cluster_id":"u1","cluster_name":"default","replica_id":"1","replica_name":"r1"}                      |      | 2023-05-18 09:03:51.502+00
(2 rows)
```

**Comment by teskje** [comment]
> Item No1 . No audit events are emitted for the migration of the system replicas when upgrading from a previous version:

Wait, we never emit audit events for system replicas! This is intentional, since we don't want to bill users for them.

> At the same time, there is a create+drop audit event for the creation of that same replica prior to the migration:

Those events are for a user replica (notice the `"cluster_id":"u1"`), not a system replica.

**Comment by philip-stoev** [comment]
> Those events are for a user replica (notice the "cluster_id":"u1"), not a system replica.

Yes, sorry.

**Review by philip-stoev — APPROVED** [review]
I just pushed a platform-checks based test that should cover this particular migration . Please take it along for the ride to main.

File: test/cloudtest/test_upgrade.py [hunk 2]
```diff
@@ -45,7 +45,7 @@ def test_upgrade(aws_region: Optional[str]) -> None:
     """Test upgrade from the last released verison to the current source by running all the Platform Checks"""
 
     mz = MaterializeApplication(tag=LAST_RELEASED_VERSION, aws_region=aws_region)
-    wait(condition="condition=Ready", resource="pod/cluster-u1-replica-1-0")
+    wait(condition="condition=Ready", resource="pod/cluster-u1-replica-u1-0")
```
Comment (philip-stoev) [inline test/cloudtest/test_upgrade.py:48]: The cloudtest upgrade is now failing -- https://buildkite.com/materialize/nightlies/builds/2402#01882e60-c710-42cd-bb33-9a3c127eddf8 . Apparently the pod name is not changed on upgrade, so possibly the old name still applies?

**Review by mjibson — APPROVED** [review]

**Review by philip-stoev — APPROVED** [review]
This PR has been tested by the QA team previously, has tests and all.

**Comment by necaris** [comment]
@teskje is the Cloud change still required?

**Comment by teskje** [comment]
> @teskje is the Cloud change still required?

Yes, the `check_environment` function still has the issue that it checks the old pod name. I was thinking, maybe instead of having a switch based on the env version we could instead wait for the pod by its `cluster-id` label instead of the pod name? That would be unaffected by this change and generally more robust.

File: src/stash/protos/objects.proto [hunk 3]
```diff
@@ -458,7 +461,7 @@ message AuditLogEventV1 {
 
     message CreateClusterReplicaV1 {
         string cluster_id = 1;
-        string cluser_name = 2;
+        string cluster_name = 2;
```
Comment (antiguru) [inline src/stash/protos/objects.proto:464]: :100:

**Review by antiguru — APPROVED** [review]
Thanks, looks good!

**Comment by teskje** [comment]
Sorry for letting this rot for so long. @antiguru and me have now rebased everything on the current main, fixed any conflicts and ported the stash migrations to the new protobuf based framework. Cloudtests are also finally passing now (I hope).

@mjibson (or someone else from @MaterializeInc/surfaces) could you give the adapter changes another pass? Especially the two stash migration PRs are completely new code that hasn't been reviewed yet.

File: src/stash/protos/objects.proto (hunk 3 above)
Comment (mjibson) [reply src/stash/protos/objects.proto:464]: Is it safe to change protobuf field names without bumping the number?

File: src/stash/src/objects.rs [hunk 4]
```diff
@@ -190,6 +190,8 @@ pub unsafe trait WireCompatible<T: prost::Message>: prost::Message + Default {
     }
 }
 
+unsafe impl WireCompatible<()> for () {}
```
Comment (mjibson) [inline src/stash/src/objects.rs:193]: What's this needed for? If it's needed, the docs say the macro should be used.

**Review by mjibson — APPROVED** [review]

**Comment by philip-stoev** [comment]
@teskje I noticed that the upgrade test on your branch reports weird warnings about version incompatibility -- https://buildkite.com/materialize/nightlies/builds/3404#annotation-018a1714-dbfc-4f47-bd91-9ba22e529fcf-error  I think this is because `latest` has now moved beyond your branch. Prior to pushing, can you please do one final rebase to make sure that the upgrade test is clean?

File: src/stash/protos/objects.proto (hunk 3 above)
Comment (teskje) [reply src/stash/protos/objects.proto:464]: Yes, names and orders of fields can change without affecting backwards-compatibility. The important thing is that the tag value remains unchanged.

File: src/stash/src/objects.rs (hunk 4 above)
Comment (teskje) [reply src/stash/src/objects.rs:193]: It is needed for `migrate_compat` to work on the `AUDIT_LOG_COLLECTION`:
```rust
AUDIT_LOG_COLLECTION.migrate_compat(tx, |_| updates).await?;
```

This collection has a value type of `()`. So without the `WireCompatible` impl, you get this error:
```
error[E0277]: the trait bound `(): WireCompatible<()>` is not satisfied
   --> src/stash/src/upgrade/v33_to_v34.rs:214:26
    |
214 |     AUDIT_LOG_COLLECTION.migrate_compat(tx, |_| updates).await?;
    |                          ^^^^^^^^^^^^^^ the trait `WireCompatible<()>` is not implemented for `()`
    |
    = help: the following other types implement trait `WireCompatible<T>`:
              <objects_v27::DatabaseKey as WireCompatible<v27_to_v28::objects_v28::DatabaseKey>>
              <objects_v27::DatabaseValue as WireCompatible<v27_to_v28::objects_v28::DatabaseValue>>
              <objects_v27::DefaultPrivilegesKey as WireCompatible<v27_to_v28::objects_v28::DefaultPrivilegesKey>>
              <objects_v27::DefaultPrivilegesValue as WireCompatible<v27_to_v28::objects_v28::DefaultPrivilegesValue>>
              <objects_v27::SchemaKey as WireCompatible<v27_to_v28::objects_v28::SchemaKey>>
              <objects_v27::SchemaValue as WireCompatible<v27_to_v28::objects_v28::SchemaValue>>
              <objects_v33::RoleKey as WireCompatible<v32_to_v33::objects_v32::RoleKey>>
              <objects_v33::RoleValue as WireCompatible<v32_to_v33::objects_v32::RoleValue>>
            and 16 others
note: required by a bound in `upgrade::<impl TypedCollection<K, V>>::migrate_compat`
   --> src/stash/src/upgrade.rs:148:20
    |
141 |     pub(crate) async fn migrate_compat<K2, V2>(
    |                         -------------- required by a bound in this associated function
...
148 |         V2: Data + WireCompatible<V>,
    |                    ^^^^^^^^^^^^^^^^^ required by this bound in `upgrade::<impl TypedCollection<K, V>>::migrate_compat`
```

Using the `wire_compatible!` macro doesn't work because it expects its arguments to be `ident`s, which `()` is not. I'm sure we could change the definition of the macro, but I would rather do that as a separate PR.

Comment (teskje) [reply src/stash/src/objects.rs:193]: https://github.com/MaterializeInc/materialize/pull/21323
